[oneTBB] Add adaptive mutexes (#354)
oneapi-src/oneAPI-spec: source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst

> .. SPDX-FileCopyrightText: 2020-2021 Intel Corporation

Reviewer: ```suggestion
.. SPDX-FileCopyrightText: 2021 Intel Corporation
```

..
.. SPDX-License-Identifier: CC-BY-4.0

=====
mutex
=====
**[mutex.mutex]**

A ``mutex`` is a class that models the :doc:`Mutex requirement <../named_requirements/mutexes/mutex>`,

> using adaptive approach: the combination of spinlock and waiting on system primitives.

Author: ```suggestion
using an adaptive approach, which is a combination of spinlock and waiting on system primitives.
```

> using an adaptive approach, which is a combination of spinlock and waiting on system primitives.

Author: ```suggestion
using an adaptive approach, it guarantees: that the thread that can't acquire the lock spins before blocking.
```

> using an adaptive approach, it guarantees: that the thread that can't acquire the lock spins before blocking.

Reviewer 2: ```suggestion
using an adaptive approach, it guarantees that the thread that cannot acquire the lock spins before blocking.
```

.. code:: cpp

    // Defined in header <oneapi/tbb/mutex.h>

    namespace oneapi {
        namespace tbb {
            class mutex {
            public:
                mutex() noexcept;
                ~mutex();

                mutex(const mutex&) = delete;
                mutex& operator=(const mutex&) = delete;

                class scoped_lock;

                void lock();
                bool try_lock();
                void unlock();

                static constexpr bool is_rw_mutex = false;
                static constexpr bool is_recursive_mutex = false;
                static constexpr bool is_fair_mutex = false;
            };
        }
    }

Member classes
--------------

.. namespace:: oneapi::tbb::mutex
	       
.. cpp:class:: scoped_lock

    The corresponding ``scoped_lock`` class. See :doc:`Mutex requirement <../named_requirements/mutexes/mutex>`.

Member functions
----------------

.. cpp:function:: mutex()

    Constructs a ``mutex`` with the unlocked state.

--------------------------------------------------

.. cpp:function:: ~mutex()

    Destroys an unlocked ``mutex``.

--------------------------------------------------

.. cpp:function:: void lock()

>     Acquires a lock. It uses adaptive logic for waiting, thus it is blocked after a certain time of busy waiting.

Reviewer 2: ```suggestion
    Acquires a lock. It uses an adaptive logic for waiting, thus it is blocked after a certain time of busy waiting.
```
